README.md
# dserve

Tool to serve a collection of datasets over HTTP.

## Running

To test:

    export FLASK_APP=scripts/server.py
    flask run


CHANGELOG.rst
CHANGELOG
=========

This project uses `semantic versioning <http://semver.org/>`_.
This change log uses principles from `keep a changelog <http://keepachangelog.com/>`_.


[Unreleased]
~~~~~~~~~~~~

Added
^^^^^


Changed
^^^^^^^

- Updated code to work with ``dtoolcore<=0.14.0``


Deprecated
^^^^^^^^^^


Removed
^^^^^^^


Fixed
^^^^^


Security
^^^^^^^^



tests/test_functional.py
"""Functional tests for dserve server."""

import os
import subprocess
import shutil
import time
import tempfile

import pytest
import requests

HERE = os.path.dirname(__file__)
APP = os.path.join(HERE, "..", "dserve", "__init__.py")
SAMPLE_DATASET_PATH = os.path.join(HERE, "data", "cotyledon_images")

TEST_SERVER = "http://127.0.0.1:{}"


@pytest.fixture(scope="module")
def run_server(request):
    port = "5001"
    server = subprocess.Popen(["python", APP, SAMPLE_DATASET_PATH, "-p", port])
    time.sleep(1)

    @request.addfinalizer
    def teardown():
        server.terminate()
    return TEST_SERVER.format(port)


@pytest.fixture()
def run_write_server(request):
    d = tempfile.mkdtemp()
    tmp_dataset_path = os.path.join(d, "cotyledon_images")
    shutil.copytree(SAMPLE_DATASET_PATH, tmp_dataset_path)

    port = "5002"
    server = subprocess.Popen(["python", APP, tmp_dataset_path, "-p", port])
    time.sleep(1)

    @request.addfinalizer
    def teardown():
        server.terminate()
        shutil.rmtree(d)

    return TEST_SERVER.format(port)


def test_root_route(run_server):
    url = run_server
    r = requests.get(url)
    assert r.status_code == 200
    assert r.headers['content-type'].find("json") != -1
    assert r.headers['access-control-allow-origin'] == "*"

    expected_content = {
        "_links": {
            "self": {"href": "/"},
            "items": {"href": "/items"},
            "overlays": {"href": "/overlays"}
        },
        "uuid": "4396cc68-2ded-4ca6-9992-bd24c11f5e09",
        "dtool_version": "0.12.1",
        "name": "cotyledon_images",
        "creator_username": "olssont",
    }
    assert r.json() == expected_content


def test_items_route(run_server):
    url = "/".join([run_server, "items"])
    r = requests.get(url)
    assert r.status_code == 200
    assert r.headers['content-type'].find("json") != -1
    assert r.headers['access-control-allow-origin'] == "*"

    expected_content = {
        "_links": {
            "self": {"href": "/items"},
        },
        "_embedded": {
            "items": [{
                "_links": {"self": {"href": "/items/290d3f1a902c452ce1c184ed793b1d6b83b59164"}},  # NOQA
                "identifier": "290d3f1a902c452ce1c184ed793b1d6b83b59164",
            }, {
                "_links": {"self": {"href": "/items/09648d19e11f0b20e5473594fc278afbede3c9a4"}},  # NOQA
                "identifier": "09648d19e11f0b20e5473594fc278afbede3c9a4",
            }],
        }
    }
    assert r.json() == expected_content


def test_overlays_route(run_server):
    url = "/".join([run_server, "overlays"])
    r = requests.get(url)
    assert r.status_code == 200
    assert r.headers['content-type'].find("json") != -1
    assert r.headers['access-control-allow-origin'] == "*"

    expected_content = {
        "_links": {
            "self": {"href": "/overlays"},
            "coordinates": {"href": "/overlays/coordinates"},
            "hash": {"href": "/overlays/hash"},
            "size": {"href": "/overlays/size"},
            "path": {"href": "/overlays/path"},
            "mimetype": {"href": "/overlays/mimetype"},
            "mtime": {"href": "/overlays/mtime"}
        },
    }
    assert r.json() == expected_content


def test_specific_item_route(run_server):
    url = "/".join([
        run_server,
        "items",
        "290d3f1a902c452ce1c184ed793b1d6b83b59164"])
    r = requests.get(url)
    assert r.status_code == 200
    assert r.headers['content-type'].find("json") != -1
    assert r.headers['access-control-allow-origin'] == "*"

    expected_content = {
        "_links": {
            "self": {"href": "/items/290d3f1a902c452ce1c184ed793b1d6b83b59164"},  # NOQA
            "content": {"href": "/items/290d3f1a902c452ce1c184ed793b1d6b83b59164/raw"},  # NOQA
            "overlays": {"href": "/items/290d3f1a902c452ce1c184ed793b1d6b83b59164/overlays"}  # NOQA
        },
        "coordinates": {"x": 4.0, "y": 5.6},
        "hash": "290d3f1a902c452ce1c184ed793b1d6b83b59164",
        "mimetype": "image/png",
        "mtime": 1491989058.0,
        "path": "first_image.png",
        "size": 276
    }
    assert r.json() == expected_content


def test_nonexisitng_specific_item_route(run_server):
    url = "/".join([
        run_server,
        "items",
        "nonsense"])
    r = requests.get(url)
    assert r.status_code == 404


def test_specific_item_raw_route(run_server):
    url = "/".join([
        run_server,
        "items",
        "290d3f1a902c452ce1c184ed793b1d6b83b59164",
        "raw"])
    r = requests.get(url)
    assert r.status_code == 200
    assert r.headers['content-type'].find("png") != -1
    assert r.headers['access-control-allow-origin'] == "*"
    assert int(r.headers['content-length']) == 276


def test_nonexisting_specific_item_raw_route(run_server):
    url = "/".join([
        run_server,
        "items",
        "nonsense",
        "raw"])
    r = requests.get(url)
    assert r.status_code == 404


def test_specific_item_overlays_route(run_server):
    url = "/".join([
        run_server,
        "items",
        "290d3f1a902c452ce1c184ed793b1d6b83b59164",
        "overlays"])
    r = requests.get(url)
    assert r.status_code == 200
    assert r.headers['content-type'].find("json") != -1
    assert r.headers['access-control-allow-origin'] == "*"

    expected_content = {
        "_links": {
            "self": {"href": "/items/290d3f1a902c452ce1c184ed793b1d6b83b59164/overlays"},  # NOQA
            "coordinates": {"href": "/overlays/coordinates/290d3f1a902c452ce1c184ed793b1d6b83b59164"},
            "hash": {"href": "/overlays/hash/290d3f1a902c452ce1c184ed793b1d6b83b59164"},
            "size": {"href": "/overlays/size/290d3f1a902c452ce1c184ed793b1d6b83b59164"},
            "path": {"href": "/overlays/path/290d3f1a902c452ce1c184ed793b1d6b83b59164"},
            "mimetype": {"href": "/overlays/mimetype/290d3f1a902c452ce1c184ed793b1d6b83b59164"},
            "mtime": {"href": "/overlays/mtime/290d3f1a902c452ce1c184ed793b1d6b83b59164"}
        }
    }
    assert r.json() == expected_content


def test_nonexisting_specific_item_overlays_route(run_server):
    url = "/".join([
        run_server,
        "items",
        "nonsense",
        "overlays"])
    r = requests.get(url)
    assert r.status_code == 404


def test_specific_item_overlay_route(run_server):
    url = "/".join([
        run_server,
        "overlays",
        "hash",
        "290d3f1a902c452ce1c184ed793b1d6b83b59164"])
    r = requests.get(url)
    assert r.status_code == 200
    assert r.headers['content-type'].find("json") != -1
    assert r.headers['access-control-allow-origin'] == "*"

    expected_content = "290d3f1a902c452ce1c184ed793b1d6b83b59164"
    assert r.json() == expected_content


def test_specific_item_nonexisting_overlay_route(run_server):
    url = "/".join([
        run_server,
        "overlays",
        "nonsense",
        "290d3f1a902c452ce1c184ed793b1d6b83b59164"])
    r = requests.get(url)
    assert r.status_code == 404


def test_specific_overlay_route(run_server):
    url = "/".join([run_server, "overlays", "coordinates"])
    r = requests.get(url)
    assert r.status_code == 200
    assert r.headers['content-type'].find("json") != -1
    assert r.headers['access-control-allow-origin'] == "*"

    expected_content = {
        "290d3f1a902c452ce1c184ed793b1d6b83b59164": {"x": 4.0, "y": 5.6},
        "09648d19e11f0b20e5473594fc278afbede3c9a4": {"x": 80.8, "y": 3.3}
    }
    assert r.json() == expected_content


def test_nonexisting_specific_overlay_route(run_server):
    url = "/".join([run_server, "overlays", "nonsese"])
    r = requests.get(url)
    assert r.status_code == 404


def test_create_new_overlay(run_write_server):
    url = "/".join([run_write_server, "overlays", "results"])
    r = requests.put(url)
    assert r.status_code == 201

    r = requests.get(url)
    assert r.status_code == 200

    expected_content = {
        "290d3f1a902c452ce1c184ed793b1d6b83b59164": {},
        "09648d19e11f0b20e5473594fc278afbede3c9a4": {}
    }
    assert r.json() == expected_content


def test_create_new_overlay_does_not_overwrite(run_write_server):
    url = "/".join([run_write_server, "overlays", "results"])
    r = requests.put(url)
    assert r.status_code == 201
    r = requests.put(url)
    assert r.status_code == 409


def test_update_specific_item_in_overlay(run_write_server):
    url = "/".join([
        run_write_server,
        "overlays",
        "coordinates",
        "290d3f1a902c452ce1c184ed793b1d6b83b59164"])
    r = requests.get(url)
    assert r.status_code == 200
    assert r.headers['content-type'].find("json") != -1
    assert r.headers['access-control-allow-origin'] == "*"

    expected_content = {"x": 4.0, "y": 5.6}
    assert r.json() == expected_content

    # application/json content-type not specified.
    r = requests.put(url, data={"x": 10.0, "y": -7.0})
    assert r.status_code == 422

    # application/json content-type specified, but data is not json.
    r = requests.put(
        url,
        data={"x": 10.0, "y": -7.0},
        headers={'content-type': 'application/json'})
    assert r.status_code == 400

    r = requests.put(
        url,
        data='{"x": 10.0, "y": -7.0}',
        headers={'content-type': 'application/json'})
    assert r.status_code == 201

    r = requests.get(url)
    expected_content = {"x": 10.0, "y": -7.0}
    assert r.json() == expected_content


def test_update_specific_item_in_readonly_overlay(run_write_server):
    url = "/".join([
        run_write_server,
        "overlays",
        "hash",
        "290d3f1a902c452ce1c184ed793b1d6b83b59164"])
    r = requests.put(
        url,
        data='"alternative_identifier"',
        headers={'content-type': 'application/json'})
    assert r.status_code == 405


def test_update_specific_item_in_nonexisting_overlay_404(run_write_server):
    url = "/".join([
        run_write_server,
        "items",
        "290d3f1a902c452ce1c184ed793b1d6b83b59164",
        "nonexisting"])

    r = requests.put(
        url,
        data='{"x": 10.0, "y": -7.0}',
        headers={'content-type': 'application/json'})
    assert r.status_code == 404


def test_update_specific_nonsese_item_in_overlay_404(run_write_server):
    url = "/".join([
        run_write_server,
        "items",
        "nonsense",
        "coordinates"])

    r = requests.put(
        url,
        data='{"x": 10.0, "y": -7.0}',
        headers={'content-type': 'application/json'})
    assert r.status_code == 404


scripts/server.py
"""Simple dataset server."""

import os
import argparse

from flask import (
    Flask,
    send_file,
    jsonify,
    request,
)

from flask_cors import CORS, cross_origin

from dtool import DataSet, NotDtoolObject

app = Flask(__name__)
cors = CORS(app)
app.config['CORS_HEADERS'] = 'Content-Type'

DATA_ROOT = '/Users/hartleym/data_repo'


@app.route('/datasets')
def show_datasets():
    all_datasets = find_all_datasets(DATA_ROOT)

    dataset_names = list(all_datasets.keys())

    return jsonify(dataset_names)


@app.route('/dataset/<dataset_name>')
def show_dataset_metadata(dataset_name):

    dataset_path = os.path.join(DATA_ROOT, dataset_name)

    dataset = DataSet.from_path(dataset_path)

    return jsonify(dataset.descriptive_metadata)


@app.route('/dataset/<dataset_name>/hashes')
def show_dataset_hashes(dataset_name):

    dataset_path = os.path.join(DATA_ROOT, dataset_name)

    dataset = DataSet.from_path(dataset_path)

    file_list = dataset.manifest["file_list"]

    hashes = [item['hash'] for item in file_list]

    return jsonify(hashes)


@app.route('/dataset/<dataset_name>/item/<item_hash>')
@cross_origin()
def server_dataset_item(dataset_name, item_hash):

    dataset_path = os.path.join(DATA_ROOT, dataset_name)
    dataset = DataSet.from_path(dataset_path)
    # FIXME - BROKEN FUNCTION
    # item_path = dataset.item_path_from_hash(item_hash)

    items_by_hash = {item['hash']: item
                     for item in dataset.manifest["file_list"]}

    item_path = os.path.join(
        dataset._abs_path,
        dataset.data_directory,
        items_by_hash[item_hash]['path']
    )
    data_mimetype = items_by_hash[item_hash]['mimetype']

    return send_file(item_path, data_mimetype)


@app.route('/dataset/<dataset_name>/item/<item_hash>/coords')
def find_item_overlay(dataset_name, item_hash):

    dataset_path = os.path.join(DATA_ROOT, dataset_name)
    dataset = DataSet.from_path(dataset_path)

    coords = dataset.overlays["coords"][item_hash]

    print(coords)

    return jsonify(coords)


@app.route('/dataset/<dataset_name>/item/<item_hash>/update_coords',
           methods=['POST'])
@cross_origin()
def set_item_coords(dataset_name, item_hash):

    dataset_path = os.path.join(DATA_ROOT, dataset_name)
    dataset = DataSet.from_path(dataset_path)

    try:
        coords = dataset.overlays["coords"]
    except KeyError:
        coords = dataset.empty_overlay()

    print('req', request.data)

    new_coords_value = request.get_json()

    coords[item_hash] = new_coords_value

    dataset.persist_overlay("coords", coords, overwrite=True)

    print(new_coords_value)

    return item_hash


def identifiers_where_overlay_is_true(dataset, overlay_name):

    overlays = dataset.overlays
    selected = [identifier
                for identifier in dataset.identifiers
                if overlays[overlay_name][identifier]]

    return selected


@app.route('/givemejpegurls', methods=['GET'])
@cross_origin()
def get_all_jpegs():

    dataset_name = 'cad_tilling_2016'
    dataset_path = os.path.join(DATA_ROOT, dataset_name)
    dataset = DataSet.from_path(dataset_path)

    jpeg_identifiers = identifiers_where_overlay_is_true(dataset, "is_jpeg")

    return jsonify(jpeg_identifiers)


def find_all_datasets(path):
    """Return dictionary of datasets in given path, keyed by name of dataset.
    Names are expected to be unique.

    :param path: Path to directory containing datasets.
    :returns: Dictionary of datasets.
    """

    directories = [d
                   for d in os.listdir(path)
                   if os.path.isdir(os.path.join(path, d))]

    all_datasets = {}
    for d in directories:
        try:
            dataset = DataSet.from_path(os.path.join(path, d))
            all_datasets[dataset.name] = dataset
        except NotDtoolObject:
            pass

    return all_datasets


def main():
    parser = argparse.ArgumentParser(description=__doc__)

    parser.add_argument('path', help='Path to data collectiont to server')

    args = parser.parse_args()

    print(find_all_datasets(args.path))


if __name__ == '__main__':
    app.run(debug=True)


dserve/__init__.py
"""Script for running the dserve server."""

import os
import argparse

from flask import (
    Flask,
    jsonify,
    send_file,
    abort,
    request,
)
from flask_cors import CORS, cross_origin

from dtoolcore import DataSet

app = Flask(__name__)
cors = CORS(app)


@app.route("/")
@cross_origin()
def root():
    content = {
        "_links": {
            "self": {"href": "/"},
            "items": {"href": "/items"},
            "overlays": {"href": "/overlays"}
        },
        "uuid": app._dataset._admin_metadata["uuid"],
        "dtool_version": app._dataset._admin_metadata["dtool_version"],
        "name": app._dataset._admin_metadata["name"],
        "creator_username": app._dataset._admin_metadata["creator_username"],
    }
    return jsonify(content)


def items_root():
    items = []
    for i in app._dataset.manifest["file_list"]:
        item = {
            "_links": {"self": {"href": "/items/{}".format(i["hash"])}},
            "identifier": i["hash"],
        }
        items.append(item)

    content = {
        "_links": {
            "self": {"href": "/items"},
        },
        "_embedded": {
            "items": items,
        }
    }
    return jsonify(content)


def specific_item(identifier):
    try:
        item = app._dataset.item_from_identifier(identifier)
    except KeyError:
        abort(404)
    content = {
        "_links": {
            "self": {"href": "/items/{}".format(identifier)},
            "content": {"href": "/items/{}/raw".format(identifier)},
            "overlays": {"href": "/items/{}/overlays".format(identifier)},
        },
    }

    overlays = app._dataset.access_overlays()
    for overlay_name, overlay in overlays.items():
        content[overlay_name] = overlay[identifier]

    return jsonify(content)


@app.route("/items")
@app.route("/items/<identifier>")
@cross_origin()
def items(identifier=None):
    if identifier is None:
        return items_root()
    else:
        return specific_item(identifier)


@app.route("/items/<identifier>/raw")
@cross_origin()
def raw_item(identifier):
    try:
        item = app._dataset.item_from_identifier(identifier)
    except KeyError:
        abort(404)
    item_path = os.path.join(
        dataset._abs_path,
        dataset.data_directory,
        item["path"]
    )
    return send_file(item_path, item["mimetype"])


@app.route("/items/<identifier>/overlays")
@cross_origin()
def item_overlays(identifier):
    try:
        item = app._dataset.item_from_identifier(identifier)
    except KeyError:
        abort(404)
    content = {
        "_links": {
            "self": {"href": "/items/{}/overlays".format(identifier)},
        },
    }
    overlays = app._dataset.access_overlays()
    for overlay_name in overlays.keys():
        href = "/overlays/{}/{}".format(overlay_name, identifier)
        content["_links"][overlay_name] = {"href": href}

    return jsonify(content)


@app.route("/overlays/<overlay>/<identifier>", methods=["GET", "PUT"])
@cross_origin()
def item_overlay_content(overlay, identifier):
    overlays = app._dataset.access_overlays()
    try:
        requested_overlay = overlays[overlay]
        requested_overlay[identifier]
    except KeyError:
        abort(404)

    if request.method == "PUT":
        if not request.is_json:
            abort(422)
        new_value = request.get_json()
        requested_overlay[identifier] = new_value
        try:
            app._dataset.persist_overlay(
                overlay, requested_overlay, overwrite=True)
        except KeyError:
            abort(405)
        return "", 201
    elif request.method == "GET":
        value = requested_overlay[identifier]
        return jsonify(value)


def overlay_root():
    overlays = app._dataset.access_overlays()
    content = {
        "_links": {
            "self": {"href": "/overlays"}},
    }
    for overlay_name in overlays.keys():
        value = {"href": "/overlays/{}".format(overlay_name)}
        content["_links"][overlay_name] = value
    return jsonify(content)


def specific_overlay(overlay_name):
    overlays = app._dataset.access_overlays()
    try:
        overlay = overlays[overlay_name]
    except KeyError:
        abort(404)
    return jsonify(overlay)


def creaate_new_overlay(overlay_name):
    empty_overlay = app._dataset.empty_overlay()
    try:
        app._dataset.persist_overlay(overlay_name, empty_overlay)
    except IOError:
        abort(409)
    return "", 201


@app.route("/overlays")
@app.route("/overlays/<overlay_name>", methods=["GET", "PUT"])
@cross_origin()
def overalys(overlay_name=None):
    if overlay_name is None:
        return overlay_root()
    else:
        if request.method == "PUT":
            return creaate_new_overlay(overlay_name)
        elif request.method == "GET":
            return specific_overlay(overlay_name)


def main(dataset, port, debug):
    app._dataset = dataset
    app.run(port=port, debug=debug)


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("dataset_path")
    parser.add_argument("-p", "--port", type=int, default=5000)
    parser.add_argument("--debug", action="store_true")
    args = parser.parse_args()
    dataset = DataSet.from_path(args.dataset_path)
    main(dataset, args.port, args.debug)
